readme.md

```diff
@@ -1,41 +1,102 @@
+to process entities.
+
+In Drupal 8, it is recommended that all processing on fields be done at the 
+entity level. As such, the service primarily takes entities as inputs and 
+processes those fields in order to either decrypt encrypted fields or encrypt 
+fields before storage.
+
+The `encrypt_entity()` and `decrypt_entity()` methods perform these actions. 
+Additionally, after we have changed the storage settings (enabled / disabled
+ encryption), we must process existing fields. `encrypt_stored_field()` and 
+ `decrypt_stored_field()` provide that functionality.
+
+Inside the service, we iterate over each field and then each of the fields 
+values. For example, the `text_with_summary` field type has a `value` and a 
+`summary` value. The encryption itself is then handled by other services. 
+For example, text encryption is handled by the `encryption` service as part of
+the `encrypt` module.
+
+The mapping of field types to services is done with our plugin system.
+
+### FieldEncryptionProvider Plugin System
+
+Plugins can be defined to map field types to an encryption / decryption service.
+Plugins can be created by using the provided 
+`Plugin/FieldEncryptionProvider/CoreStrings.php` plugin as a reference. 
+This plugin defines the supported fields in its annotation, and loads the 
+encryption service used to process these strings.
+By using the plugin system, we allow other modules to define encryption 
+processes for non-string values such as numbers, files, and images. 
+These plugins can also define encryption process for string values of custom 
+field types.
+
+### Encryption Services
+The field_encrypt module uses services to encrypt and decrypt field values. 
+The module relies on the `encrypt` module for string processing. Other modules 
+may define their own services and map fields to those services with
+ `FieldEncryptionProvider` plugins. They should extend the 
+ `FieldEncryptionProviderBase` class.
+
+### Encrypt Field Storage Third Party Setting
+In Drupal 8, the field storage settings (field base in Drupal 7) are stored in
+ configuration management using a Configuration Entity. 
+ Extending configuration entities is done by providing `Third Party Settings` 
+ and using the associated methods.
+
+We provide this setting to all fields using the 
+`config/schema/field_encrypt.schema.yml` file.
+
+### Configuring Fields to use Encryption
+While this creates the setting, we need to modify the field storage form so that
+we can set / change this value. We hook into the form system with 
+`hook_form_alter()` in our `.module` file.
+
+### Updating Stored Field Values
+In addition to adding this new value to the form, we add a function to handle 
+the form submission which will respond to changes in the setting value.
+This way, when the setting is changed, we can process stored values and 
+encrypt / decrypt to match the new setting. This encryption / decryption is 
```

Comment:
Perfect - I didn't see this represented but this speaks to my data integrity concern for changing field settings.